### PR TITLE
MYST-101 client. list identities teqilapi

### PR DESCRIPTION
**Pull request by shroomist**
I needed IdentityManager and KeystoreManager to be public

### CODE REVIEW

**Comment by Waldz** [comment]
1. Use PR format, Jira relates it to ticket
`StoryID Explanation what was changed`
e.g. "MYST-101 Client. API method "list identities""

2. Dont need to mention that PR contains tests, because tests is standard anyway

File: cmd/mysterium_client/command_run/command.go [hunk 1]
```diff
@@ -3,16 +3,19 @@ package command_run
 import (
 	"encoding/json"
 	"errors"
+	"io"
```
Comment (Waldz) [inline cmd/mysterium_client/command_run/command.go]: We should agree on code formatting guidelines.
Otherwise it will start to reformat it forward&backward.

File: cmd/mysterium_client/command_run/command.go [hunk 2]
```diff
@@ -71,7 +74,9 @@ func (cmd *CommandRun) Run(options CommandOptions) (err error) {
 		return err
 	}
 
-	apiEndpoints := tequilapi.NewApiEndpoints()
+	idm := identity.NewIdentityManager("TEST")
```
Comment (Waldz) [inline cmd/mysterium_client/command_run/command.go]: This password should be input for identity create method, e.g. it's optional for listing identities

File: cmd/mysterium_client/command_run/command.go [hunk 3]
```diff
@@ -71,7 +74,9 @@ func (cmd *CommandRun) Run(options CommandOptions) (err error) {
...
+	router := tequilapi.NewApiEndpoints()
+	router.GET("/list-identities", endpoints.ListIdentitiesEndpointFactory(idm).ListIdentities)
```
Comment (Waldz) [inline cmd/mysterium_client/command_run/command.go]: Lets reconsider this, as main controller will start to be fat

Comment (Waldz) [inline cmd/mysterium_client/command_run/command.go]: Read on RESTful resources naming conventions:
- https://restfulapi.net/resource-naming/
- http://www.restapitutorial.com/lessons/restfulresourcenaming.html

File: identity/manager.go [hunk 4]
```diff
@@ -26,23 +26,23 @@ func accountToIdentity(account accounts.Account) *dto.Identity {
 	return &identity
 }
 
-func identityToAccount(identityString string) accounts.Account {
+func IdentityToAccount(identityString string) accounts.Account {
```
Comment (Waldz) [inline identity/manager.go]: This method was intentionally private

File: identity/manager.go [hunk 5]
```diff
@@ -11,13 +11,13 @@ import (
 	"strings"
 )
 
-type identityManager struct {
-	keystoreManager keystoreInterface
+type IdentityManager struct {
+	KeystoreManager keystoreInterface
```
Comment (Waldz) [inline identity/manager.go]: This attribute was intentionally private.
- lets force usage of factory `NewIdentityManager()`
- lets ramake to `NewIdentityManager(keydir string)` -> `NewIdentityManager(keystoryManager)`

File: tequilapi/endpoints/list_identities.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+	"github.com/mysterium/node/service_discovery/dto"
+	"github.com/mysterium/node/tequilapi/utils"
+)
+
+type identityList struct {
+	Identities []dto.Identity `json:"identities"`
+}
+
+type listIdentitiesEndpoint struct {
+	listIdentities func() []dto.Identity
```
Comment (Waldz) [inline tequilapi/endpoints/list_identities.go]: I would inject `IdentityManager` as dependency

File: tequilapi/endpoints/list_identities.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+func ListIdentitiesEndpointFactory(idm *identity.IdentityManager) *listIdentitiesEndpoint {
+	return &listIdentitiesEndpoint{listIdentities: idm.GetIdentities}
+}
+
+func (lsid *listIdentitiesEndpoint) ListIdentities(writer http.ResponseWriter, request *http.Request, params httprouter.Params) {
+	list := identityList{
+		Identities: lsid.listIdentities(),
```
Comment (Waldz) [inline tequilapi/endpoints/list_identities.go]: I miss DTO layer which is mapping internal structure -> to API structures

File: tequilapi/endpoints/list_identities_test.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+            "identities":["0x000000000000000000000000000000000000000A"]
+        }`,
+		resp.Body.String())
+}
+
+func newManager(accountValue string) *identity.IdentityManager {
+	return &identity.IdentityManager{
+		KeystoreManager: &identity.KeyStoreFake{
+			AccountsMock: []accounts.Account{
+				identity.IdentityToAccount(accountValue),
```
Comment (Waldz) [inline tequilapi/endpoints/list_identities_test.go]: Dont call internal function of package

File: tequilapi/endpoints/list_identities_test.go [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+	req := httptest.NewRequest("GET", "/irrelevant", nil)
+	resp := httptest.NewRecorder()
+
+	mockIdm := newManager("0x000000000000000000000000000000000000000A")
+	handlerFunc := ListIdentitiesEndpointFactory(mockIdm).ListIdentities
+	handlerFunc(resp, req, httprouter.Params{})
+
+	assert.JSONEq(
+		t,
+		`{
```
Comment (Waldz) [inline tequilapi/endpoints/list_identities_test.go]: Looks like misformated

File: tequilapi/endpoints/list_identities_test.go [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+            "identities":["0x000000000000000000000000000000000000000A"]
```
Comment (Waldz) [inline tequilapi/endpoints/list_identities_test.go]: Lets have identity as structure, because it will have more fields later:
```
{
    "id": "0x000000000000000000000000000000000000000A"
}
```

File: tequilapi/endpoints/list_identities_test.go [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+func TestListIdentities(t *testing.T) {
+
+	req := httptest.NewRequest("GET", "/irrelevant", nil)
+	resp := httptest.NewRecorder()
+
+	mockIdm := newManager("0x000000000000000000000000000000000000000A")
+	handlerFunc := ListIdentitiesEndpointFactory(mockIdm).ListIdentities
```
Comment (Waldz) [inline tequilapi/endpoints/list_identities_test.go]: Factory could return `Handler` callback

File: tequilapi/endpoints/list_identities_test.go [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+	handlerFunc(resp, req, httprouter.Params{})
```
Comment (Waldz) [inline tequilapi/endpoints/list_identities_test.go]: I thing 3rd parameter could be optional. Should try

File: cmd/mysterium_client/command_run/command.go (hunk 3 above)
Comment (shroomist) [reply cmd/mysterium_client/command_run/command.go]: changed to '/identities'

File: identity/manager.go (hunk 4 above)
Comment (shroomist) [reply identity/manager.go]: ok, made it private again

File: identity/manager.go (hunk 5 above)
Comment (shroomist) [reply identity/manager.go]: done

File: tequilapi/endpoints/list_identities.go (hunk 6 above)
Comment (shroomist) [reply tequilapi/endpoints/list_identities.go]: now injected on IdentityHandlers construction.
or are you suggesting to inject it for every api call separately?
we now only have `list` but there might be others..

File: tequilapi/endpoints/list_identities.go (hunk 7 above)
Comment (shroomist) [reply tequilapi/endpoints/list_identities.go]: please check `service_discovery/dto/identity.go` if it's what you meant to say

File: tequilapi/endpoints/list_identities_test.go (hunk 8 above)
Comment (shroomist) [reply tequilapi/endpoints/list_identities_test.go]: fixed

File: tequilapi/endpoints/list_identities_test.go (hunk 9 above)
Comment (shroomist) [reply tequilapi/endpoints/list_identities_test.go]: fixed

File: tequilapi/endpoints/list_identities_test.go (hunk 10 above)
Comment (shroomist) [reply tequilapi/endpoints/list_identities_test.go]: done

File: tequilapi/endpoints/list_identities_test.go (hunk 11 above)
Comment (shroomist) [reply tequilapi/endpoints/list_identities_test.go]: factory now returns an object with `Get` handler.
I think we might have a pattern where we'd have a single object for each endpoint with `Get` `Put` `Delete` handlers
let me know if you still want a factory that returns single callbacks for each endpoint method

File: tequilapi/endpoints/list_identities_test.go (hunk 12 above)
Comment (shroomist) [reply tequilapi/endpoints/list_identities_test.go]: it's not optional in this router library

File: tequilapi/endpoints/list_identities.go (hunk 7 above)
Comment (Waldz) [reply tequilapi/endpoints/list_identities.go]: REST DTOs should be nest to REST service. See example of Healthcheck endpoint

File: tequilapi/interface.go [hunk 13]
```diff
@@ -0,0 +1,10 @@
+package tequilapi
+
+import (
+	"github.com/julienschmidt/httprouter"
+	"net/http"
+)
+
+type CollectionTequilaInterface interface {
```
Comment (Waldz) [inline tequilapi/interface.go]: What is das?

File: tequilapi/endpoints/identities.go [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+	"github.com/mysterium/node/tequilapi/utils"
+)
+
+type IdentitiesApi struct {
+	getIdentityArray func() []dto.Identity
+}
+
+func IdentityHandlers(idm *identity.IdentityManager) *IdentitiesApi {
+	return &IdentitiesApi{getIdentityArray: idm.GetIdentities}
+}
```
Comment (tadovas) [inline tequilapi/endpoints/identities.go]: Why do we need intermediate getIdentityArray func if we depend on IdentityManager anyway? just store the reference in the struct and call GetIdentities when needed.
Another way would be get rid of IdentityManager dependecy and pass getIdentityArray to IdentityHandlers. But current situation is overabstracted and doesnt add anything

File: identity/controller.go [hunk 15]
```diff
@@ -1,17 +1,17 @@
...
 
 type identityHandler struct {
-	manager *identityManager
+	manager *IdentityManager
 	cache   *identityCache
 }
 
-func SelectIdentity(dir string, nodeKey string) (id *dto.Identity, err error) {
-	handler := NewIdentityHandler(dir)
+func SelectIdentity(keystore keystoreInterface, nodeKey string) (id *dto.Identity, err error) {
```
Comment (tadovas) [inline identity/controller.go]: passing key store to select identity looks bad, shouldn't this function be part of identitymanager? I hope it's in the middle of refactoring

Comment (shroomist) [reply identity/controller.go]: @ignasbernotas is taking care of this from here afaik.

File: tequilapi/endpoints/identities.go (hunk 14 above)
Comment (shroomist) [reply tequilapi/endpoints/identities.go]: fixed

**Comment by shroomist** [comment]
I would like to merge this.
`Test_CreateNewIdentity` is failing, but some voodoomagic happening there. when merged will create a ticket to fix this test, or submit your fix. @interro
@tadovas your comments addressed.
 @Waldz waiting for your approval and merging tonight.

**Comment by tadovas** [comment]
Failing tests == failing builds. I would like to have all tests fixed/refactored before accepting pull request

**Comment by shroomist** [comment]
tests fixed

File: tequilapi/http_routes_test.go [hunk 16]
```diff
@@ -19,7 +20,8 @@ func (hfts *handleFunctionTestStruct) httprouterHandle(resp http.ResponseWriter,
 func TestHttpRouterHandlesRequests(t *testing.T) {
 	ts := handleFunctionTestStruct{false}
 
-	router := NewApiEndpoints()
+	idmFake := identity.NewIdentityManagerFake()
+	router := NewApiEndpoints(idmFake)
```
Comment (Waldz) [inline tequilapi/http_routes_test.go]: You dont need to boo all API routes, while testing only one Router.
Construct `Router{}` structure without factory

File: tequilapi/endpoints/identities_test.go [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+import (
+	"net/http/httptest"
+	"testing"
+
+	"github.com/julienschmidt/httprouter"
+	"github.com/mysterium/node/identity"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestListIdentities(t *testing.T) {
```
Comment (Waldz) [inline tequilapi/endpoints/identities_test.go]: What about one more test then zero identities exist?

File: tequilapi/endpoints/identities.go [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+import (
+	"net/http"
+
+	"github.com/julienschmidt/httprouter"
+	"github.com/mysterium/node/identity"
+	"github.com/mysterium/node/service_discovery/dto"
+	"github.com/mysterium/node/tequilapi/utils"
+)
+
+type IdentityData struct {
```
Comment (Waldz) [inline tequilapi/endpoints/identities.go]: - `IdentityData` -> `IdentityDto`
- `Identities ` -> `IdentityListDto`
- DTO structures does not need to be public

File: identity/keystore_fake.go [hunk 19]
```diff
@@ -5,42 +5,40 @@ import (
 	"github.com/ethereum/go-ethereum/common"
 )
 
-type keyStoreFake struct {
+type KeyStoreFake struct {
```
Comment (Waldz) [inline identity/keystore_fake.go]: - `KeyStoreFake` does not need to be public, it's interface is public

File: identity/manager.go [hunk 20]
```diff
@@ -32,8 +31,8 @@ func identityToAccount(identityString string) accounts.Account {
 	}
 }
 
-func (idm *identityManager) CreateNewIdentity(passphrase string) (*dto.Identity, error) {
-	account, err := idm.keystoreManager.NewAccount(passphrase)
+func (idm *identityManager) CreateNewIdentity(addrToHex string) (*dto.Identity, error) {
+	account, err := idm.keystoreManager.NewAccount(addrToHex)
```
Comment (Waldz) [inline identity/manager.go]: Should be `passphrase`, doublecheck

File: tequilapi/endpoints/identities_test.go [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+			},
+			{
+				"id": "0x000000000000000000000000000000000000000F"
+			}]
+		}`,
+		resp.Body.String())
+}
+
+func newManager() identity.IdentityManagerInterface {
+	keystoreFake := &identity.KeyStoreFake{}
```
Comment (Waldz) [inline tequilapi/endpoints/identities_test.go]: Use your fake identity manager

File: tequilapi/endpoints/identities.go [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+}
+type Identities struct {
+	Identities []IdentityData `json:"identities"`
+}
+
+type IdentitiesApi struct {
+	idm identity.IdentityManagerInterface
+}
+
+func NewIdentitiesEndpoint(idm identity.IdentityManagerInterface) *IdentitiesApi {
```
Comment (Waldz) [inline tequilapi/endpoints/identities.go]: `IdentitiesApi` does not need to be private

File: cmd/mysterium_client/command_run/command.go [hunk 23]
```diff
@@ -71,10 +74,12 @@ func (cmd *CommandRun) Run(options CommandOptions) (err error) {
 		return err
 	}
 
-	apiEndpoints := tequilapi.NewApiEndpoints()
-	//TODO additional endpoint registration can go here i.e apiEndpoints.GET("/path", httprouter.Handle function)
+	// options.keystoreDir still to be implemented. represents keystore directory/file
+	keystore := keystore.NewKeyStore(options.DirectoryKeystore, keystore.StandardScryptN, keystore.StandardScryptP)
```
Comment (ignasbernotas) [inline cmd/mysterium_client/command_run/command.go]: Variable 'keystore' collides with imported package name

Comment (shroomist) [reply cmd/mysterium_client/command_run/command.go]: nice catch

File: identity/manager_test.go [hunk 24]
```diff
@@ -27,12 +28,13 @@ func newManagerWithError(errorMock error) *identityManager {
...
-	identity, err := manager.CreateNewIdentity("")
+	manager := NewIdentityManagerFake()
+	assert.Len(t, manager.ksmFake.Accounts(), 0)
+	identity, err := manager.CreateNewIdentity("0x000000000000000000000000000000000000bEEf")
 
 	assert.NoError(t, err)
-	assert.Equal(t, *identity, dto.Identity("0x000000000000000000000000000000000000bEEF"))
-	assert.Len(t, manager.keystoreManager.Accounts(), 2)
+	assert.Equal(t, dto.Identity("0x000000000000000000000000000000000000bEEf"), *identity)
+	assert.Len(t, manager.ksmFake.Accounts(), 1)
```
Comment (Waldz) [inline identity/manager_test.go]: Dont adopt test by failing result

File: identity/keystore_fake.go [hunk 25]
```diff
@@ -7,20 +7,24 @@ import (
...
+func NewKeystoreFake() *keyStoreFake {
+    return &keyStoreFake{}
 }
 
 func (self *keyStoreFake) Accounts() []accounts.Account {
 	return self.AccountsMock
 }
 
-func (self *keyStoreFake) NewAccount(passphrase string) (accounts.Account, error) {
+func (self *keyStoreFake) NewAccount(address string) (accounts.Account, error) {
```
Comment (Waldz) [inline identity/keystore_fake.go]: It's passphrase, not address

File: identity/keystore_fake.go [hunk 26]
```diff
@@ -7,20 +7,24 @@ import (
...
 	if self.ErrorMock != nil {
 		return accounts.Account{}, self.ErrorMock
 	}
 
 	accountNew := accounts.Account{
-		Address: common.HexToAddress("0x000000000000000000000000000000000000bEEF"),
+		Address: common.HexToAddress(address),
```
Comment (Waldz) [inline identity/keystore_fake.go]: All new accounts was create with ending "...bEEF"

File: identity/manager_fake.go [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+	"github.com/mysterium/node/service_discovery/dto"
+)
+
+type idmFake struct {
+    ksmFake keystoreInterface
+}
+
+func NewIdentityManagerFake() *idmFake {
+	return &idmFake{
+	    ksmFake:NewKeystoreFake(),
```
Comment (Waldz) [inline identity/manager_fake.go]: Dont depend on keystore, then You dont need it
